chainstate/tx-verifier/src/transaction_verifier/input_output_policy/tests.rs

```diff
@@ -0,0 +1,721 @@
+
+    let result = check_tx_inputs_outputs_purposes(&tx, &utxo_db).unwrap_err();
+    assert_eq!(result, ConnectTransactionError::InvalidOutputTypeInTx);
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn tx_many_to_one(#[case] seed: Seed) {
+    let check = |inputs_utxos: &[TxOutput], source_outputs: &[TxOutput], expected_result| {
+        let mut rng = make_seedable_rng(seed);
+        let number_of_inputs = rng.gen_range(2..10);
+        let (utxo_db, tx) = prepare_utxos_and_tx_with_random_combinations(
+            &mut rng,
+            inputs_utxos,
+            number_of_inputs,
+            source_outputs,
+            1,
+            None,
+        );
+        assert_eq!(
+            check_tx_inputs_outputs_purposes(&tx, &utxo_db),
+            expected_result
+        );
+    };
+
+    // valid cases
+    let valid_inputs = [lock_then_transfer(), transfer()];
+    let valid_outputs = [
+        lock_then_transfer(),
+        transfer(),
+        burn(),
+        stake_pool(),
+        create_delegation(),
+        delegate_staking(),
+    ];
+    check(&valid_inputs, &valid_outputs, Ok(()));
+
+    // invalid outputs
+    let invalid_outputs = [produce_block()];
+    check(
+        &valid_inputs,
+        &invalid_outputs,
+        Err(ConnectTransactionError::InvalidOutputTypeInTx),
+    );
+}
+
+#[rstest]
+#[trace]
+#[case(Seed::from_entropy())]
+fn tx_many_to_many(#[case] seed: Seed) {
+    let check = |source_inputs: &[TxOutput], source_outputs: &[TxOutput], expected_result| {
+        let mut rng = make_seedable_rng(seed);
+        let number_of_inputs = rng.gen_range(2..10);
+        let number_of_outputs = rng.gen_range(2..10);
+
+        let (utxo_db, tx) = prepare_utxos_and_tx_with_random_combinations(
+            &mut rng,
+            source_inputs,
+            number_of_inputs,
+            source_outputs,
+            number_of_outputs,
+            None,
+        );
+        assert_eq!(
+            check_tx_inputs_outputs_purposes(&tx, &utxo_db),
+            expected_result
+        );
+    };
+
+    // valid cases
+    let valid_inputs = [lock_then_transfer(), transfer()];
+    let valid_outputs = [lock_then_transfer(), transfer(), burn(), delegate_staking()];
+    check(&valid_inputs, &valid_outputs, Ok(()));
+
+    // invalid cases
+    let invalid_inputs = [burn(), stake_pool(), produce_block(), create_delegation()];
+    let invalid_outputs = [stake_pool(), produce_block(), create_delegation()];
+    check(
+        &invalid_inputs,
+        &valid_outputs,
+        Err(ConnectTransactionError::InvalidInputTypeInTx),
+    );
+    check(
+        &valid_inputs,
+        &invalid_outputs,
+        Err(ConnectTransactionError::InvalidOutputTypeInTx),
+    );
+}
+
+#[rstest]
+    #[rustfmt::skip]
+    #[case(transfer(), transfer(),           Err(ConnectTransactionError::InvalidInputTypeInReward))]
+    #[case(transfer(), burn(),               Err(ConnectTransactionError::InvalidInputTypeInReward))]
```

Comment:
Extra indentation again.

Btw I just find it's possible to instruct rustfmt to only skip formatting of certain attributes. May be worth giving a shot. https://github.com/rust-lang/rustfmt#tips

Reply:
Unfortunately "custom inner attributes are unstable"

Reply:
Oh, that's a shame. Thanks for checking.